pkg/agent/openflow/pipeline.go

```diff
@@ -768,71 +837,73 @@ func (c *client) l2ForwardCalcFlow(dstMAC net.HardwareAddr, ofPort uint32, skipI
-// Agent after L2forwarding calculation.
+// traceflowL2ForwardOutputFlows generates Traceflow specific flows that outputs traceflow packets
+// to OVS port and Antrea Agent after L2forwarding calculation.
 func (c *client) traceflowL2ForwardOutputFlows(dataplaneTag uint8, category cookie.Category) []binding.Flow {
 	flows := []binding.Flow{}
-	// Output and SendToController if output port is tunnel or gateway port.
-	// The gw0 IP as Traceflow destination is not supported.
-	if c.encapMode.SupportsEncap() {
-		flows = append(flows, c.pipeline[L2ForwardingOutTable].BuildFlow(priorityNormal+3).
-			MatchReg(int(PortCacheReg), config.DefaultTunOFPort).
-			MatchIPDscp(dataplaneTag).
-			SetHardTimeout(300).
-			MatchProtocol(binding.ProtocolIP).
-			MatchRegRange(int(marksReg), portFoundMark, ofPortMarkRange).
-			Action().OutputRegRange(int(PortCacheReg), ofPortRegRange).
-			Action().SendToController(uint8(PacketInReasonTF)).
-			Cookie(c.cookieAllocator.Request(category).Raw()).
-			Done())
-		flows = append(flows, c.pipeline[L2ForwardingOutTable].BuildFlow(priorityNormal+3).
-			MatchReg(int(PortCacheReg), config.DefaultTunOFPort).
+	for _, ipProtocol := range []binding.Protocol{binding.ProtocolIP, binding.ProtocolIPv6} {
+		if c.encapMode.SupportsEncap() {
+			// SendToController and Output if output port is tunnel port.
+			flows = append(flows, c.pipeline[L2ForwardingOutTable].BuildFlow(priorityNormal+3).
+				MatchReg(int(PortCacheReg), config.DefaultTunOFPort).
+				MatchIPDscp(dataplaneTag).
+				SetHardTimeout(300).
+				MatchProtocol(ipProtocol).
+				MatchRegRange(int(marksReg), portFoundMark, ofPortMarkRange).
+				Action().OutputRegRange(int(PortCacheReg), ofPortRegRange).
+				Action().SendToController(uint8(PacketInReasonTF)).
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Done())
+			// Only SendToController if output port is local gateway.
+			flows = append(flows, c.pipeline[L2ForwardingOutTable].BuildFlow(priorityNormal+2).
+				MatchReg(int(PortCacheReg), config.HostGatewayOFPort).
+				MatchIPDscp(dataplaneTag).
+				SetHardTimeout(300).
+				MatchProtocol(ipProtocol).
+				MatchRegRange(int(marksReg), portFoundMark, ofPortMarkRange).
+				Action().SendToController(uint8(PacketInReasonTF)).
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Done())
+		} else {
+			// SendToController and Output if output port is local gateway.
+			flows = append(flows, c.pipeline[L2ForwardingOutTable].BuildFlow(priorityNormal+2).
+				MatchReg(int(PortCacheReg), config.HostGatewayOFPort).
+				MatchIPDscp(dataplaneTag).
+				SetHardTimeout(300).
+				MatchProtocol(ipProtocol).
+				MatchRegRange(int(marksReg), portFoundMark, ofPortMarkRange).
+				Action().OutputRegRange(int(PortCacheReg), ofPortRegRange).
+				Action().SendToController(uint8(PacketInReasonTF)).
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Done())
```

Comment:
It looks like in encap mode the packet will be output when the destination IP is external while it won't be output in noencap mode? Why is the difference?

Reply:
I kind of get it. The latter case could be packets sent to Pods on other Nodes.

Reply:
Correct, let me add a comment there